### PR TITLE
feat: support formatting SAS code

### LINKED ISSUE #28: Format SAS code
The Syntax Highlighting and Pop-up Syntax are very great and usefull.
I wonder if there is the functinality of formating SAS code like the SAS EG or SAS Studio.

### PR DESCRIPTION
**Pull request by scnwwu**
**Summary**
Resolve #28

**Testing**

1. Open a `.sas` file
2. Open context menu and select `Format Document`
3. See formatted SAS code

### CODE REVIEW

File: server/src/sas/formatter/parser.ts [hunk 1]
```diff
@@ -0,0 +1,304 @@
...
+  children: (Region | Statement)[];
+}
+
+export type SASAST = Program | Region | Statement | Token;
+
+const isComment = (token: Token) =>
+  token.type === "comment" || token.type === "macro-comment";
+
+const getPreStatement = (parent: Program | Region) => {
+  let preStatement = parent.children[parent.children.length - 2];
```
Comment (Enzhihpp) [inline server/src/sas/formatter/parser.ts:37]: whether there are cases where parent.children.length is less than 2?

Comment (scnwwu) [reply server/src/sas/formatter/parser.ts]: Refactored to keep a reference for previous statement

File: server/src/sas/LanguageServiceProvider.ts [hunk 2]
```diff
@@ -54,6 +55,7 @@ export class LanguageServiceProvider {
   private syntaxProvider;
   public completionProvider;
   public formatOnTypeProvider;
+  public formatter;
```
Comment (scottdover) [inline server/src/sas/LanguageServiceProvider.ts:58]: Does this need to be `public`?

File: server/src/sas/formatter/parser.ts [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+      if (!currentStatement) {
+        currentStatement = {
+          type: "statement",
+          name: node.text,
+          children: [],
+        };
+        if (region) {
+          if (region.children.length === 0) {
+            // section start
+            if (prevStatement) {
```
Comment (scottdover) [inline server/src/sas/formatter/parser.ts:162]:
```suggestion
          if (region.children.length === 0 && prevStatement) {
```
Looks like this can _maybe_ be combined

File: server/src/sas/formatter/printer.ts [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+  return node.leadingComment
+    ? [node.leadingComment.text, hardline, statement]
+    : statement;
+};
+
+const removeRedundantLineBreaks = (docs: Doc[]) =>
+  docs.map((doc, index) => {
+    if (index < docs.length - 1) {
+      const nextDoc = docs[index + 1];
+      if (startsWithLineBreak(nextDoc)) {
```
Comment (scottdover) [inline server/src/sas/formatter/printer.ts:94]:
```suggestion
    if (index < docs.length - 1 && startsWithLineBreak(docs[index + 1])) {
```

**Review by scottdover — COMMENTED** [review]
This is _a little_ hard to follow. There's a huge amount of code within `getParser`. I'm wondering if there is any way to improve readibility there (split it into functions that perform smaller operations, add more comments. There's also quite a few things like...

```
if (region && region.block) {
  const block = region.block;
  if (
    block.endLine === node.end.line &&
    block.endCol === node.end.column
  ) {
    region = parents.pop();
  }
}
```

that could be rewritten as a single condition/fewer indentations, like

```
if (
  region &&
  region.block &&
  region.block.endLine === node.end.line &&
  region.block.endCol === node.end.column
) {
  region = parents.pop();
}
```

File: server/src/sas/formatter/parser.ts [hunk 5]
```diff
@@ -0,0 +1,303 @@
...
+
+const hasParent = (parents: Region[], block: FoldingBlock) => {
+  let index = parents.length - 1;
+  while (index >= 0 && !parents[index].block) {
+    index--;
+  }
+  const parent = parents[index]?.block;
+  if (!parent) {
+    return false;
+  }
```
[the remark this answers is not in the document]
Comment (scnwwu) [reply server/src/sas/formatter/parser.ts]: done in bd4862a

File: server/src/sas/formatter/parser.ts (hunk 3 above)
Comment (scnwwu) [reply server/src/sas/formatter/parser.ts]: done in bd4862a

File: server/src/sas/formatter/printer.ts (hunk 4 above)
Comment (scnwwu) [reply server/src/sas/formatter/printer.ts]: done in bd4862a

**Review by scnwwu — COMMENTED** [review]
Made some improvement. Thank you.

**Review by boyce-w — APPROVED** [review]

**Comment by Zhirong2022** [comment]
Acceptance Criteria:

1.Make sure there are no regression issues 
2.Verify that code indent/outdent works as expected after applying formatting
2.1 Code indent/outdent in Region
2.2 Code indent/outdent in recursive Region
2.3 Code indent/outdent in GTL
2.4 Code indent/outdent in DS2
3.Verify that each statement will hold a line if having many statements written in one line 
4.Verity that a long statement can be wrapped properly 
5.Verify that unnecessary whitespaces will be trimmed in the code
6.Verify that datalines will be preserved as expected
7.Verify that datalines4 will be preserved as expected
8.Verify that the leading comment will be positioned properly, it associates with the statement next to it 
9.Verify that the trailing comment will be positioned properly, it associates with the statement on its left
10.Verify that do/end pair in the code will be formatted as expected
11.Verify that equal sign in the code can be handled properly 
12.Verify that the embedded language Python will be preserved well
13.Verify that the embedded language Lua will be preserved well
14.Verify 1 to 13 in OS: Windows and Mac
15.Verify 1 to 14 in NLS environment

**Comment by Zhirong2022** [comment]
It does not format quite good with the code below.
```
* define the ds2 package thru which the fcmp functions will be called;
proc ds2;
package pkg /
   overwrite=yes
   language='fcmp'
   table='base.fcmpsubs';
run;

* call fcmp thru the ds2 wrapper package;
data;
   dcl package pkg p();
   dcl double a b;
   method init();
      do a = 10 to 20;
         b=p.square(a);
         put a= b=;
      end;
   end;
enddata;
run;
quit;
```

**Comment by scnwwu** [comment]
> It does not format quite good with the code below.
> 
> ```
> * define the ds2 package thru which the fcmp functions will be called;
> proc ds2;
> package pkg /
>    overwrite=yes
>    language='fcmp'
>    table='base.fcmpsubs';
> run;
> 
> * call fcmp thru the ds2 wrapper package;
> data;
>    dcl package pkg p();
>    dcl double a b;
>    method init();
>       do a = 10 to 20;
>          b=p.square(a);
>          put a= b=;
>       end;
>    end;
> enddata;
> run;
> quit;
> ```

It relies on PACKAGE/ENDPACKAGE pairs. When ENDPACKAGE omitted, it is hard. Will leave it to future.

**Comment by Zhirong2022** [comment]
method/end is not formatted properly with the code below.
```
proc ds2;
    ds2_options sas;
    data work.Shots(keep=(ShotId text PName DPO _count));
        dcl package json j();
        method init();
            dcl package myhttp p();
            dcl int rc hstat tokenType parseFlags;
            LookForShots: do while ( rc < 101 );
            j.getNextToken( rc, token, tokenType, parseFlags );
            *put token=;
            if ( j.isLabel(tokenType, parseFlags) and (token EQ 'Shots') )
                then goto ShotsArray;
        end;
        put 'Error: "Shots" not found.';
        return;
    end; * End method init;
    enddata;
    run;
quit; * End of PROC DS2 step;

```

**Comment by scnwwu** [comment]
> method/end is not formatted properly with the code below.
> 
> ```
> proc ds2;
>     ds2_options sas;
>     data work.Shots(keep=(ShotId text PName DPO _count));
>         dcl package json j();
>         method init();
>             dcl package myhttp p();
>             dcl int rc hstat tokenType parseFlags;
>             LookForShots: do while ( rc < 101 );
>             j.getNextToken( rc, token, tokenType, parseFlags );
>             *put token=;
>             if ( j.isLabel(tokenType, parseFlags) and (token EQ 'Shots') )
>                 then goto ShotsArray;
>         end;
>         put 'Error: "Shots" not found.';
>         return;
>     end; * End method init;
>     enddata;
>     run;
> quit; * End of PROC DS2 step;
> ```

Fixed. Thank you

**Comment by Zhirong2022** [comment]
It looks better if the wrapped comments align with it's upper. Suggest to make some improvement.
```
proc ds2;
    data _null_;
        dcl decimal(30) id;
        dcl package sqlstmt fsqlclass('select id from mylib.class where name1= trim(?)',
            [name]);

        /* Define a return code variable. */
        method run();
            dcl int rc;

            /* Execute the Fsqlclass package instance on the in-memory data to apply the TRIM function
  and WHERE clause on column Name. Capture any error codes. */
            rc=fsqlclass.execute();
            if (rc) then put 'ERROR: Error executing';

            /* Bind the ID column to the results returned by Fsqlclass, 
 capturing any return codes. */
            rc=fsqlclass.bindResults([id]);
            if (rc) then put 'ERROR: Error binding results';

            /* Fetch the results and print them to the SAS log. */
            rc=fsqlclass.fetch();
            if (rc=0) then put name= id=;
        end;
    enddata;
    run;
quit;

```

**Comment by scnwwu** [comment]
> It looks better if the wrapped comments align with it's upper. Suggest to make some improvement.
> 
> ```
> proc ds2;
>     data _null_;
>         dcl decimal(30) id;
>         dcl package sqlstmt fsqlclass('select id from mylib.class where name1= trim(?)',
>             [name]);
> 
>         /* Define a return code variable. */
>         method run();
>             dcl int rc;
> 
>             /* Execute the Fsqlclass package instance on the in-memory data to apply the TRIM function
>   and WHERE clause on column Name. Capture any error codes. */
>             rc=fsqlclass.execute();
>             if (rc) then put 'ERROR: Error executing';
> 
>             /* Bind the ID column to the results returned by Fsqlclass, 
>  capturing any return codes. */
>             rc=fsqlclass.bindResults([id]);
>             if (rc) then put 'ERROR: Error binding results';
> 
>             /* Fetch the results and print them to the SAS log. */
>             rc=fsqlclass.fetch();
>             if (rc=0) then put name= id=;
>         end;
>     enddata;
>     run;
> quit;
> ```

Fixed. Thank you

File: client/test/languageServer/formatter.test.ts [hunk 6]
```diff
@@ -0,0 +1,25 @@
+import * as vscode from "vscode";
+
+import * as assert from "assert";
```
Comment (scottdover) [inline client/test/languageServer/formatter.test.ts:3]: I notice `chai` has an assert function as well. Why do we use this one over the chai version?

**Review by scottdover — APPROVED** [review]

File: client/test/languageServer/formatter.test.ts (hunk 6 above)
Comment (scnwwu) [reply client/test/languageServer/formatter.test.ts:3]: The "assert" is a built-in module. I would not add dependency until there's a need.

**Comment by Zhirong2022** [comment]
noprint will hold a line unexpectedly.
```
  proc kde data=sashelp.heart;
   bivar height(ngrid=8) weight(ngrid=10) / out=kde(keep=value1 value2 count) noprint plots=none;
  run;

```

**Comment by scnwwu** [comment]
> noprint will hold a line unexpectedly.
> 
> ```
>   proc kde data=sashelp.heart;
>    bivar height(ngrid=8) weight(ngrid=10) / out=kde(keep=value1 value2 count) noprint plots=none;
>   run;
> ```

Fixed. Thank you.

**Comment by Zhirong2022** [comment]
Error "Request textDocument/formatting failed." is thrown if trying to format the code below.
```
%a
proc print data=sashelp.cars;run;

```

**Comment by scnwwu** [comment]
> Error "Request textDocument/formatting failed." is thrown if trying to format the code below.
> 
> ```
> %a
> proc print data=sashelp.cars;run;
> ```

Fixed. Thank you

**Comment by Zhirong2022** [comment]
It will wrap to a new line if the parameter ends with semicolon.  
```
%let _execenv=%nrquote(%nrstr(SASStudio;));
```
image

**Comment by Zhirong2022** [comment]
Lua code is not preserved well after applying formatting
```
proc lua;
submit;
   if (sas.exists("work.homes")) then
      local t = sas.read_ds("work.homes")
      i=1
      while(t[i] ~= nil) do
         print("Obs #" .. i)
         for k,v in pairs(t[i]) do 
             print(k,v) 
         end
         print("\n")
         i = i+1
      end
   end
endsubmit;
run;
```

**Comment by Zhirong2022** [comment]
Lua code is not preserved well after applying formatting
```
proc lua;

submit;

   local rc

   local code = [[
         data sample; set answer;
         where CCUID = @ccuid@;
         y = @subValue@;
         run;
   ]]

   rc = sas.submit(code, {ccuid="67", subValue=72})

endsubmit;
run;

```

**Comment by scnwwu** [comment]
All issues reported above fixed. Thanks.

**Comment by Zhirong2022** [comment]
All issues have been fixed.

**Review by Zhirong2022 — APPROVED** [review]
